<a href="https://colab.research.google.com/github/Pragna235/ML-Assignments/blob/main/Modified_Credit_Data_Analysis_Task_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing Modules and Libraries

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import sys

# Reading the Dataset


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
dataset = pd.read_csv("/content/drive/MyDrive/Credit-Modified.csv")
dataset

Unnamed: 0   Income  Limit  Rating  Cards  Age  Education  Gender-num  \
0             1   14.891   3606     283      2   34         11           0   
1             2  106.025   6645     483      3   82         15           1   
2             3  104.593   7075     514      4   71         11           0   
3             4  148.924   9504     681      3   36         11           1   
4             5   55.882   4897     357      2   68         16           0   
..          ...      ...    ...     ...    ...  ...        ...         ...   
395         396   12.096   4100     307      3   32         13           0   
396         397   13.364   3838     296      5   65         17           0   
397         398   57.872   4171     321      5   67         12           1   
398         399   37.728   2525     192      1   44         13           0   
399         400   18.701   5524     415      5   64          7           1   

     Gender  Student-num Student Married         Ethnicity  Balance  \
0      Male            0      No     Yes         Caucasian      333   
1    Female            1     Yes     Yes             Asian      903   
2      Male            0      No      No             Asian      580   
3    Female            0      No      No             Asian      964   
4      Male            0      No     Yes         Caucasian      331   
..      ...          ...     ...     ...               ...      ...   
395    Male            0      No     Yes         Caucasian      560   
396    Male            0      No      No  African American      480   
397  Female            0      No     Yes         Caucasian      138   
398    Male            0      No     Yes         Caucasian        0   
399  Female            0      No      No             Asian      966   

     Defaultee dcat  
0            0   No  
1            0   No  
2            0   No  
3            0   No  
4            0   No  
..         ...  ...  
395          0   No  
396          0   No  
397          0   No  
398          0   No  
399          0   No  

[400 rows x 16 columns]

# Dataset Analysis


In [4]:
# Print first 5 rows of the dataset
dataset.head()

Unnamed: 0   Income  Limit  Rating  Cards  Age  Education  Gender-num  \
0           1   14.891   3606     283      2   34         11           0   
1           2  106.025   6645     483      3   82         15           1   
2           3  104.593   7075     514      4   71         11           0   
3           4  148.924   9504     681      3   36         11           1   
4           5   55.882   4897     357      2   68         16           0   

   Gender  Student-num Student Married  Ethnicity  Balance  Defaultee dcat  
0    Male            0      No     Yes  Caucasian      333          0   No  
1  Female            1     Yes     Yes      Asian      903          0   No  
2    Male            0      No      No      Asian      580          0   No  
3  Female            0      No      No      Asian      964          0   No  
4    Male            0      No     Yes  Caucasian      331          0   No

In [5]:
# Print last 5 rows of the dataset
dataset.tail()

Unnamed: 0  Income  Limit  Rating  Cards  Age  Education  Gender-num  \
395         396  12.096   4100     307      3   32         13           0   
396         397  13.364   3838     296      5   65         17           0   
397         398  57.872   4171     321      5   67         12           1   
398         399  37.728   2525     192      1   44         13           0   
399         400  18.701   5524     415      5   64          7           1   

     Gender  Student-num Student Married         Ethnicity  Balance  \
395    Male            0      No     Yes         Caucasian      560   
396    Male            0      No      No  African American      480   
397  Female            0      No     Yes         Caucasian      138   
398    Male            0      No     Yes         Caucasian        0   
399  Female            0      No      No             Asian      966   

     Defaultee dcat  
395          0   No  
396          0   No  
397          0   No  
398          0   No  
399          0   No

In [6]:
# Print the number of rows and columns in the dataset
dataset.shape

(400, 16)

# Checking for missing and duplicate values...


In [8]:
dataset.isnull().sum()

Unnamed: 0     0
Income         0
Limit          0
Rating         0
Cards          0
Age            0
Education      0
Gender-num     0
Gender         0
Student-num    0
Student        0
Married        0
Ethnicity      0
Balance        0
Defaultee      0
dcat           0
dtype: int64

There are no missing values in the dataset.

In [10]:
dataset[dataset.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, Income, Limit, Rating, Cards, Age, Education, Gender-num, Gender, Student-num, Student, Married, Ethnicity, Balance, Defaultee, dcat]
Index: []

There are no duplicate rows.

# Conversion of Categorical and Numerical Values

In [11]:
dataset.describe()

Unnamed: 0      Income         Limit      Rating       Cards  \
count  400.000000  400.000000    400.000000  400.000000  400.000000   
mean   200.500000   45.218885   4735.600000  354.940000    2.957500   
std    115.614301   35.244273   2308.198848  154.724143    1.371275   
min      1.000000   10.354000    855.000000   93.000000    1.000000   
25%    100.750000   21.007250   3088.000000  247.250000    2.000000   
50%    200.500000   33.115500   4622.500000  344.000000    3.000000   
75%    300.250000   57.470750   5872.750000  437.250000    4.000000   
max    400.000000  186.634000  13913.000000  982.000000    9.000000   

              Age   Education  Gender-num  Student-num      Balance  \
count  400.000000  400.000000  400.000000   400.000000   400.000000   
mean    55.667500   13.450000    0.517500     0.100000   520.015000   
std     17.249807    3.125207    0.500319     0.300376   459.758877   
min     23.000000    5.000000    0.000000     0.000000     0.000000   
25%     41.750000   11.000000    0.000000     0.000000    68.750000   
50%     56.000000   14.000000    1.000000     0.000000   459.500000   
75%     70.000000   16.000000    1.000000     0.000000   863.000000   
max     98.000000   20.000000    1.000000     1.000000  1999.000000   

        Defaultee  
count  400.000000  
mean     0.180000  
std      0.384669  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      0.000000  
max      1.000000

In [12]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   400 non-null    int64  
 1   Income       400 non-null    float64
 2   Limit        400 non-null    int64  
 3   Rating       400 non-null    int64  
 4   Cards        400 non-null    int64  
 5   Age          400 non-null    int64  
 6   Education    400 non-null    int64  
 7   Gender-num   400 non-null    int64  
 8   Gender       400 non-null    object 
 9   Student-num  400 non-null    int64  
 10  Student      400 non-null    object 
 11  Married      400 non-null    object 
 12  Ethnicity    400 non-null    object 
 13  Balance      400 non-null    int64  
 14  Defaultee    400 non-null    int64  
 15  dcat         400 non-null    object 
dtypes: float64(1), int64(10), object(5)
memory usage: 50.1+ KB


In [13]:
dataset.groupby('Defaultee').size()

Defaultee
0    328
1     72
dtype: int64

In [14]:
dataset['dcat'].value_counts()

No     328
Yes     72
Name: dcat, dtype: int64

In [15]:
dataset = dataset.drop('Unnamed: 0',axis=1)
dataset = dataset.drop('dcat',axis=1)
datasetd

Income  Limit  Rating  Cards  Age  Education  Gender-num  Gender  \
0     14.891   3606     283      2   34         11           0    Male   
1    106.025   6645     483      3   82         15           1  Female   
2    104.593   7075     514      4   71         11           0    Male   
3    148.924   9504     681      3   36         11           1  Female   
4     55.882   4897     357      2   68         16           0    Male   
..       ...    ...     ...    ...  ...        ...         ...     ...   
395   12.096   4100     307      3   32         13           0    Male   
396   13.364   3838     296      5   65         17           0    Male   
397   57.872   4171     321      5   67         12           1  Female   
398   37.728   2525     192      1   44         13           0    Male   
399   18.701   5524     415      5   64          7           1  Female   

     Student-num Student Married         Ethnicity  Balance  Defaultee  
0              0      No     Yes         Caucasian      333          0  
1              1     Yes     Yes             Asian      903          0  
2              0      No      No             Asian      580          0  
3              0      No      No             Asian      964          0  
4              0      No     Yes         Caucasian      331          0  
..           ...     ...     ...               ...      ...        ...  
395            0      No     Yes         Caucasian      560          0  
396            0      No      No  African American      480          0  
397            0      No     Yes         Caucasian      138          0  
398            0      No     Yes         Caucasian        0          0  
399            0      No      No             Asian      966          0  

[400 rows x 14 columns]

# Label Encoding

In [16]:
from sklearn.preprocessing import LabelEncoder

label = LabelEncoder()
dataset['Gender'] = label.fit_transform(dataset['Gender'])
dataset['Student'] = label.fit_transform(dataset['Student'])
dataset['Married'] = label.fit_transform(dataset['Married'])
dataset['Ethnicity'] = label.fit_transform(dataset['Ethnicity'])

dataset

Income  Limit  Rating  Cards  Age  Education  Gender-num  Gender  \
0     14.891   3606     283      2   34         11           0       1   
1    106.025   6645     483      3   82         15           1       0   
2    104.593   7075     514      4   71         11           0       1   
3    148.924   9504     681      3   36         11           1       0   
4     55.882   4897     357      2   68         16           0       1   
..       ...    ...     ...    ...  ...        ...         ...     ...   
395   12.096   4100     307      3   32         13           0       1   
396   13.364   3838     296      5   65         17           0       1   
397   57.872   4171     321      5   67         12           1       0   
398   37.728   2525     192      1   44         13           0       1   
399   18.701   5524     415      5   64          7           1       0   

     Student-num  Student  Married  Ethnicity  Balance  Defaultee  
0              0        0        1          2      333          0  
1              1        1        1          1      903          0  
2              0        0        0          1      580          0  
3              0        0        0          1      964          0  
4              0        0        1          2      331          0  
..           ...      ...      ...        ...      ...        ...  
395            0        0        1          2      560          0  
396            0        0        0          0      480          0  
397            0        0        1          2      138          0  
398            0        0        1          2        0          0  
399            0        0        0          1      966          0  

[400 rows x 14 columns]

In [17]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Income       400 non-null    float64
 1   Limit        400 non-null    int64  
 2   Rating       400 non-null    int64  
 3   Cards        400 non-null    int64  
 4   Age          400 non-null    int64  
 5   Education    400 non-null    int64  
 6   Gender-num   400 non-null    int64  
 7   Gender       400 non-null    int64  
 8   Student-num  400 non-null    int64  
 9   Student      400 non-null    int64  
 10  Married      400 non-null    int64  
 11  Ethnicity    400 non-null    int64  
 12  Balance      400 non-null    int64  
 13  Defaultee    400 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 43.9 KB


In [18]:
dataset['Defaultee'].value_counts()

0    328
1     72
Name: Defaultee, dtype: int64

# Model Training

In [19]:
from sklearn.model_selection import train_test_split

array = dataset.values
x = array[:,:-1]
y = array[:,-1]

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.3,random_state=42)

# Training and Testing the Model using Logistic Regression

In [20]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(x_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [21]:
y_pred = model.predict(x_test)
y_pred

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 1., 1., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       1., 0., 0., 1., 1., 1., 1., 0., 0., 1., 0., 0., 1., 1., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 1., 1., 0., 0., 0., 0., 0.,
       0., 1., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.])

In [22]:
from sklearn.metrics import classification_report, accuracy_score,confusion_matrix

accuracy_score(y_test,y_pred)

0.9583333333333334

In [23]:
confusion_matrix(y_test,y_pred)

array([[97,  4],
       [ 1, 18]])

In [24]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

         0.0       0.99      0.96      0.97       101
         1.0       0.82      0.95      0.88        19

    accuracy                           0.96       120
   macro avg       0.90      0.95      0.93       120
weighted avg       0.96      0.96      0.96       120



So, the accuracy obtained using Logistic Regression is around 95.8%

# Training and Testing the Model using Support Vector Machine

In [26]:
from sklearn.svm import SVC

model = SVC()
model.fit(x_train,y_train)

SVC()

In [27]:
y_pred = model.predict(x_test)
y_pred

array([1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0.,
       0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 1., 1., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,
       0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.])

In [28]:
accuracy_score(y_test,y_pred)

0.8833333333333333

In [29]:
confusion_matrix(y_test,y_pred)

array([[97,  4],
       [10,  9]])

In [30]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

         0.0       0.91      0.96      0.93       101
         1.0       0.69      0.47      0.56        19

    accuracy                           0.88       120
   macro avg       0.80      0.72      0.75       120
weighted avg       0.87      0.88      0.87       120



So, the accuracy obtained using Support Vector Machine is approximately 89%

Therefore, Logistic Regression worked better than Support Vector Machine for the dataset